src/core/execute.c

```diff
@@ -3419,7 +3432,31 @@ static int exec_child(
 
         /* Drop groups as early as possbile */
         if (needs_setuid) {
-                r = enforce_groups(gid, supplementary_gids, ngids);
+                bool needs_cleanup = false;
+                _cleanup_free_ gid_t *gids_to_enforce = NULL;
+                int ngids_to_enforce = 0;
+
+                if (ngids > 0 && ngids_added_by_pam > 0) {
+                        gids_to_enforce = new(gid_t, ngids + ngids_added_by_pam);
+                        if (!gids_to_enforce) {
+                                log_oom();
+                        }
+
+                        ngids_to_enforce = merge_gid_lists(supplementary_gids, ngids, gids_added_by_pam, ngids_added_by_pam, gids_to_enforce);
+                        needs_cleanup = true;
+                } else if (ngids > 0) {
+                        gids_to_enforce = supplementary_gids;
+                        ngids_to_enforce = ngids;
+                } else if (ngids_added_by_pam > 0) {
+                        gids_to_enforce = gids_added_by_pam;
+                        ngids_to_enforce = ngids_added_by_pam;
+                }
+
```

Comment:
Once `merge_gid_lists` is simplified, most of this logic can be removed and `needs_cleanup` will not be necessary. Just call `merge_gid_lists` unconditionally.